### ISSUE TITLE
Remove the md5 parameter from stage files

### ISSUE THREAD

**Issue by Witiko** [iterative/dvc#2120]
Dvc [stores the top-level `md5` field in Dvc-files](https://dvc.org/doc/user-guide/dvc-file-format), so that it can detect when a stage has changed. However, this field adds no new information, since the `md5` field can be always inferred from the rest of the Dvc-file at a negligible computational cost. The field also makes manual changes to Dvc-files more difficult than they need to be, see the example below. Therefore, it would be benefitial if this field were not added to new Dvc-files, and ignored in existing Dvc-files.

Consider the following (not completely fictional) example: I execute the `dvc run -f results.dvc -o results ./compute_results.py` command. After long computation, I realize that I forgot to specify the `models` directory as the stage output. When the `md5` field is not present, I can just run `dvc add models`, and copy the output entry for the `models` directory from `models.dvc` to `results.dvc`. Presto, done! With the `md5` field present, I need to either re-run the experiment, or dig through the output of `dvc status -v` to find out the correct value of the `md5` field.

I realize that my knowledge of Dvc is far from complete, and that the `md5` field may actually serve some useful purpose. If so, please excuse my ignorance, and know that I would be happy to learn about the true purpose of the `md5` field. If there is none, then I suggest that we should get rid of the field.

**Comment by efiop** [comment]
@Witiko You can use `dvc commit` to save dvc file without re-running. In the scenario you've described, you could simply run `dvc commit results.dvc` and it will save all the needed checksums.

**Comment by Witiko** [comment]
Thank you for the workaround. However, what is the purpose of the `md5` field?

**Comment by efiop** [comment]
@Witiko It is actually not a workaround, but rather a proper solution that was added for that specific purpose(and some others) 🙂 We've got a lot of requests about it with the same types of scenarios. The purpose of `md5` is to verify that the content of the dvc file didn't change since the last time. For example, you've had a file that looks like
```
outs:
   - path: foo
      md5: f0f0f0f0f0f0f0
```
now you add a cmd field
```
cmd: ./myscript
outs:
   - path: foo
      md5: f0f0f0f0f0f0f0
```

so it basically transforms from `dvc add foo` into `dvc run -o foo ./myscript`. How would `dvc repro` know that is has changed? WIthout top-level md5, everything seems to be in check, md5 of the foo matches. That is why we need a top-level md5 that is singnaling us if the dvc file itself has changed.
